### PR TITLE
Install sdks dynamically-part3. Node sdks installation.

### PR DESCRIPTION
**Pull request by kichalla**

### CODE REVIEW

File: src/BuildScriptGenerator/Node/VersionProviders/NodeSdkStorageVersionProvider.cs [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+        {
+        }
+
+        // To enable unit testing
+        public virtual PlatformVersionInfo GetVersionInfo()
+        {
+            if (_platformVersionInfo == null)
+            {
+                _platformVersionInfo = GetAvailableVersionsFromStorage(
+                    platformName: "nodejs",
```
Comment (arroyc) [inline src/BuildScriptGenerator/Node/VersionProviders/NodeSdkStorageVersionProvider.cs]: use toolnameconstant?

File: tests/Oryx.BuildImage.Tests/Node/NodeDynamicInstallationTest.cs [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+        public NodeDynamicInstallationTest(ITestOutputHelper output) : base(output)
+        {
+        }
+
+        [Fact]
+        public void DynamicallyInstallsNodeRuntimeAndBuilds()
+        {
+            // Arrange
+            var devPackageName = "nodemon";
```
Comment (arroyc) [inline tests/Oryx.BuildImage.Tests/Node/NodeDynamicInstallationTest.cs]: add a comment may be? that nodemon and express is present in sampleapp's package.json

File: tests/Oryx.Integration.Tests/Node/NodeDynamicInstallationTest.cs [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+    {
+        private readonly string DefaultSdksRootDir = "/tmp/oryx/nodejs";
+
+        public NodeDynamicInstallationTest(ITestOutputHelper output, TestTempDirTestFixture fixture)
+            : base(output, fixture)
+        {
+        }
+
+        [Fact]
+        public async Task CanBuildAndRunApp_AfterSettingUpEnvironmentExplicitly()
```
Comment (arroyc) [inline tests/Oryx.Integration.Tests/Node/NodeDynamicInstallationTest.cs]: I would change this test from fact to theory, may be use 12 and 10 for minimal test optimization?

File: tests/Oryx.Integration.Tests/Node/NodeDynamicInstallationTest.cs [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+                    var data = await _httpClient.GetStringAsync($"http://localhost:{hostPort}/");
+                    Assert.Contains("Say It Again", data);
+                });
+        }
+
+        [Fact]
+        public async Task CanBuildAndRunApp_UsingScriptCommand()
+        {
+            // Arrange
+            var nodeVersion = "10.1";
```
Comment (arroyc) [inline tests/Oryx.Integration.Tests/Node/NodeDynamicInstallationTest.cs]: same comment

File: src/BuildScriptGenerator/Node/VersionProviders/NodeSdkStorageVersionProvider.cs (hunk 1 above)
Comment (kichalla) [reply src/BuildScriptGenerator/Node/VersionProviders/NodeSdkStorageVersionProvider.cs]: toolname is `node` but we have `nodejs` folder. This has been a problem for a while now. I will file a cleanup item for this to make everything `node` instead

File: src/startupscriptgenerator/src/node/scriptgenerator.go [hunk 5]
```diff
@@ -27,6 +27,7 @@ type NodeStartupScriptGenerator struct {
 	RemoteDebuggingPort             string
 	UseLegacyDebugger               bool //used for node versions < 7.7
 	SkipNodeModulesExtraction       bool
+	EnableDynamicInstall            bool
```
[the remark this answers is not in the document]
Comment (kichalla) [reply src/startupscriptgenerator/src/node/scriptgenerator.go]: yes, it does not matter.

File: tests/Oryx.Integration.Tests/Python/PythonDynamicInstallationTest.cs [hunk 6]
```diff
@@ -129,7 +129,7 @@ public async Task CanBuildAndRunPythonApp_UsingScriptCommandAndSetEnvSwitch()
                 .SetEnvironmentVariable(
                     SdkStorageConstants.SdkStorageBaseUrlKeyName,
                     SdkStorageConstants.DevSdkStorageBaseUrl)
-                .AddCommand($"oryx create-script -appPath {appDir} -bindPort {ContainerPort}")
+                .AddCommand($"oryx create-script -appPath {appDir} -bindPort {ContainerPort} -enableDynamicInstall")
```
Comment (arroyc) [inline tests/Oryx.Integration.Tests/Python/PythonDynamicInstallationTest.cs]: is that the fix to current validation build failure bug in master?

File: tests/Oryx.Integration.Tests/Node/NodeDynamicInstallationTest.cs [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+    {
+        private readonly string DefaultSdksRootDir = "/tmp/oryx/nodejs";
+
+        public NodeDynamicInstallationTest(ITestOutputHelper output, TestTempDirTestFixture fixture)
+            : base(output, fixture)
+        {
+        }
+
+        [Theory]
+        [InlineData("10")]
```
Comment (arroyc) [inline tests/Oryx.Integration.Tests/Node/NodeDynamicInstallationTest.cs]: use constant

File: tests/Oryx.Integration.Tests/Node/NodeDynamicInstallationTest.cs [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+                },
+                async (hostPort) =>
+                {
+                    var data = await _httpClient.GetStringAsync($"http://localhost:{hostPort}/");
+                    Assert.Contains("Say It Again", data);
+                });
+        }
+
+        [Theory]
+        [InlineData("10")]
```
Comment (arroyc) [inline tests/Oryx.Integration.Tests/Node/NodeDynamicInstallationTest.cs]: use constant

**Comment by arroyc** [comment]
Please merge PR after a successful nightly build

File: vsts/pipelines/templates/_releaseJobTemplate.yml [hunk 9]
```diff
@@ -57,8 +57,8 @@ jobs:
     skipComponentGovernanceDetection: true
 
   steps:
-  - task: DownloadBuildArtifacts@0
-    displayName: 'Download Artifacts for release'
+  - task: DownloadPipelineArtifact@2
```
Comment (arroyc) [inline vsts/pipelines/templates/_releaseJobTemplate.yml]: nit: I would prefer to keep the task version same in a specific job/task template

File: tests/Oryx.Integration.Tests/Python/PythonDynamicInstallationTest.cs (hunk 6 above)
Comment (kichalla) [reply tests/Oryx.Integration.Tests/Python/PythonDynamicInstallationTest.cs]: This is the PR having the fix. https://github.com/microsoft/Oryx/pull/505

File: vsts/pipelines/templates/_releaseJobTemplate.yml (hunk 9 above)
Comment (kichalla) [reply vsts/pipelines/templates/_releaseJobTemplate.yml]: These are different task types actually

**Pull request merged**